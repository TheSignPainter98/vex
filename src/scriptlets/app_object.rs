use std::fmt::Display;

use allocative::Allocative;
use starlark::{
    environment::{Methods, MethodsBuilder, MethodsStatic},
    eval::Evaluator,
    starlark_module,
    values::Value,
    values::{none::NoneType, NoSerialize, ProvidesStaticType, StarlarkValue},
};
use starlark_derive::starlark_value;

use crate::{
    error::Error,
    scriptlets::{
        action::Action,
        event::EventType,
        extra_data::{HandlerDataBuilder, InvocationData},
    },
};

#[derive(Debug, PartialEq, Eq, ProvidesStaticType, NoSerialize, Allocative)]
pub struct AppObject;

impl AppObject {
    pub const NAME: &'static str = "vex";

    #[starlark_module]
    fn methods(builder: &mut MethodsBuilder) {
        fn language<'v>(
            #[starlark(this)] _this: Value<'v>,
            lang: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.language", &[Action::Initing])?;

            HandlerDataBuilder::get_from(eval.module()).set_language(lang.into());

            Ok(NoneType)
        }

        fn query<'v>(
            #[starlark(this)] _this: Value<'v>,
            query: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.query", &[Action::Initing])?;

            // TODO(kcza): attach the id in errors somewhere?
            HandlerDataBuilder::get_from(eval.module()).set_query(query.into());

            Ok(NoneType)
        }

        fn observe<'v>(
            #[starlark(this)] _this: Value<'v>,
            event: &str,
            handler: Value<'v>,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(eval, "vex.observe", &[Action::Initing])?;

            let event = event.parse()?;
            HandlerDataBuilder::get_from(eval.module()).add_observer(event, handler);

            Ok(NoneType)
        }

        fn warn<'v>(
            #[starlark(this)] _this: Value<'v>,
            _msg: &'v str,
            eval: &mut Evaluator<'v, '_>,
        ) -> anyhow::Result<NoneType> {
            AppObject::check_attr_available(
                eval,
                "vex.warn",
                &[
                    Action::Vexing(EventType::Start),
                    Action::Vexing(EventType::Match),
                    Action::Vexing(EventType::EoF),
                    Action::Vexing(EventType::End),
                ],
            )?;

            todo!();
            // Ok(NoneType)
        }
    }

    fn check_attr_available(
        eval: &Evaluator<'_, '_>,
        attr_path: &'static str,
        available_actions: &[Action],
    ) -> anyhow::Result<()> {
        let curr_action = InvocationData::get_from(eval).action();
        if !available_actions.contains(&curr_action) {
            return Err(Error::Unavailable {
                what: attr_path,
                action: curr_action,
            }
            .into());
        }
        Ok(())
    }
}

starlark::starlark_simple_value!(AppObject);
#[starlark_value(type = "vex")]
impl<'v> StarlarkValue<'v> for AppObject {
    fn get_methods() -> Option<&'static Methods> {
        static RES: MethodsStatic = MethodsStatic::new();
        RES.methods(AppObject::methods)
    }
}

impl Display for AppObject {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
        Self::NAME.fmt(f)
    }
}
