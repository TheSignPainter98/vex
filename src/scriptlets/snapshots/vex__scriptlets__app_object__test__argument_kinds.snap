---
source: src/scriptlets/app_object.rs
expression: error_messages
---
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * language.star:2, in init\n      vex.add_trigger('rust')\nerror: Found 1 extra positional argument(s) for call to add_trigger\n --> language.star:2:5\n  |\n2 |     vex.add_trigger('rust')\n  |     ^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * query-without-language.star:2, in init\n      vex.add_trigger(query='(binary_expression)')\nerror: cannot add query without specifying a language\n --> query-without-language.star:2:5\n  |\n2 |     vex.add_trigger(query='(binary_expression)')\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * observe-event.star:6, in init\n      vex.observe(on_query_match, event='query_match')\nerror: Found `event` extra named parameter(s) for call to function\n --> observe-event.star:6:5\n  |\n6 |     vex.observe(on_query_match, event='query_match')\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * observe-observer.star:6, in init\n      vex.observe('query_match', observer=on_query_match)\nerror: Found `observer` extra named parameter(s) for call to function\n --> observe-observer.star:6:5\n  |\n6 |     vex.observe('query_match', observer=on_query_match)\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * warn-message.star:9, in on_query_match\n      vex.warn(message='oh no')\nerror: Missing parameter `message` for call to warn\n --> warn-message.star:9:5\n  |\n9 |     vex.warn(message='oh no')\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * warn-at.star:9, in on_query_match\n      vex.warn('oh no', (event.captures['bin_expr'], 'bin_expr'))\nerror: Found 1 extra positional argument(s) for call to warn\n --> warn-at.star:9:5\n  |\n9 |     vex.warn('oh no', (event.captures['bin_expr'], 'bin_expr'))\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * warn-show-also.star:9, in on_query_match\n      vex.warn('oh no', [(event.captures['bin_expr'], 'bin_expr')], at=(event.captu...\nerror: Found 1 extra positional argument(s) for call to warn\n --> warn-show-also.star:9:5\n  |\n9 |     vex.warn('oh no', [(event.captures['bin_expr'], 'bin_expr')], at=(event.captures['bin_expr'], 'bin_expr'))\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
- "Traceback (most recent call last):\n  File <builtin>, in <module>\n  * warn-show-also.star:9, in on_query_match\n      vex.warn('oh no', 'extra_info', at=(event.captures['bin_expr'], 'bin_expr'))\nerror: Found 1 extra positional argument(s) for call to warn\n --> warn-show-also.star:9:5\n  |\n9 |     vex.warn('oh no', 'extra_info', at=(event.captures['bin_expr'], 'bin_expr'))\n  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  |\n"
